state/store.go

```diff
@@ -502,6 +504,33 @@ func (store dbStore) saveABCIResponses(height int64, abciResponses *tmstate.ABCI
 	return store.db.SetSync(abciResponsesKey(height), bz)
 }
 
+// SaveValidatorSet is used to save a single validator set. It is exposed so
+// that a backfill operation during state sync can populate the store with the
+// necessary amount of validator sets to verify any evidence it may encounter.
+//
+// FIXME: Usually when saving validator sets we only save them when they change
+// and use a pointer likey system instead. This doesn't work when saving them in
+// reverse order so instead we are going to save every validator set regardless
+// if they changed or not. In the future we should look to optimize this
```

Comment:
could you pass this onto the caller, check the valsetHash, if they differ call save?
(havent read the rest of the code so you may be doing this already)

Reply:
I mention this in the description of the PR that we could implement some logic to keep track of when the validator set changes and then save all the validator sets at once

Reply:
I added in the logic you mentioned now